client/src/App.tsx

```
@@ -0,0 +1,188 @@
+import React, { useEffect, useState } from 'react';
+
+type Message = {
+  author: string;
+  text: string;
+  time: string;
+};
+
+export function App() {
+  const [messages, setMessages] = useState<Message[]>([]);
+  const [text, setText] = useState('');
+  const [rooms, setRooms] = useState<string[]>([]);
+  const [currentRoom, setCurrentRoom] = useState('general');
+  const [newRoom, setNewRoom] = useState('');
+  const [socket, setSocket] = useState<WebSocket | null>(null);
+
+  // Load available rooms
+  useEffect(() => {
+    fetch('http://localhost:3000/rooms')
+      .then((res) => res.json())
+      .then((data) => setRooms(data));
+  }, []);
+
+  // Connect WebSocket whenever room changes
+  useEffect(() => {
+    const ws = new WebSocket('ws://localhost:3000');
+    setSocket(ws);
+
+    ws.onopen = () => {
+      ws.send(JSON.stringify({ type: 'join', room: currentRoom }));
+    };
+
+    ws.onmessage = (event) => {
+      const data = JSON.parse(event.data);
+
+      if (data.type === 'init') {
+        setMessages(data.messages);
+      }
+
+      if (data.type === 'message') {
+        setMessages((prev) => [...prev, data.message]);
+      }
+    };
+
+    return () => ws.close();
+  }, [currentRoom]);
+
+  // Ensure username exists
+  useEffect(() => {
+    let saved = localStorage.getItem('username');
+    if (!saved) {
+      saved = prompt('Enter a username') || 'Anonymous';
+      localStorage.setItem('username', saved);
+    }
+  }, []);
+
+  function sendMessage(e: React.FormEvent) {
+    e.preventDefault();
+    if (!socket) return;
```

Before sending a message you only check `if (!socket) return;`. That allows attempts to send when the socket exists but is not open. Check the socket state: `if (!socket || socket.readyState !== WebSocket.OPEN) { /* notify user or queue */ return; }` to avoid silent failures.